staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriter
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+)
+
+// UserProvidedDecorator represensts a user (client that uses this package)
+// provided decorator that wraps an inner http.ResponseWriter object.
+// The user provided decorator object must return the inner (decorated)
+// http.ResponseWriter object via the Unwrap function.
+type UserProvidedDecorator interface {
+	http.ResponseWriter
+
+	// Unwrap returns the inner http.ResponseWriter object associated
+	// with the user provided decorator.
+	Unwrap() http.ResponseWriter
+}
+
+// WrapForHTTP1Or2 accepts a user provided decorator of an "inner" http.responseWriter
+// object and potentially wraps it with a new http.ResponseWriter object that
+// implements http.CloseNotifier, http.Flusher, and/or http.Hijacker by
+// delegating to the user provided decorator or the inner
+// http.ResponseWriter, so that the returned http.ResponseWriter object
+// implements the same subset of those interfaces as the inner http.ResponseWriter.
+// The scope is limited to http.CloseNotifier, http.Flusher, and http.Hijacker.
+// This function does not take into account any other interfaces
+// a ResponseWriter object may implement.
```

Comment:
I think this remark is redundant with the one above it.  The earlier remark is more informative.

What is missing here is an explanation of the three cases that this function covers.

```suggestion
// This function handles the following three casses.
// - The inner ResponseWriter implements `http.CloseNotifier`, `http.Flusher`, and `http.Hijacker`.
// - The inner ResponseWriter implements `http.CloseNotifier` and `http.Flusher` but not `http.Hijacker`.
// - All the other cases collapse to this one, in which the given ResponseWriter is returned.
```